src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultDocumentContextFactory.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class DefaultDocumentContextFactory : DocumentContextFactory
+    {
+        private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly DocumentVersionCache _documentVersionCache;
+        private readonly ILogger<DefaultDocumentContextFactory> _logger;
+
+        public DefaultDocumentContextFactory(
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            DocumentVersionCache documentVersionCache,
+            ILoggerFactory loggerFactory)
+        {
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _documentVersionCache = documentVersionCache;
+            _logger = loggerFactory.CreateLogger<DefaultDocumentContextFactory>();
+        }
+
+        public override async Task<DocumentContext?> TryCreateAsync(Uri documentUri, CancellationToken cancellationToken)
+        {
+            var filePath = documentUri.GetAbsoluteOrUNCPath();
+            var documentAndVersion = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                if (_documentResolver.TryResolveDocument(filePath, out var documentSnapshot))
+                {
+                    if (_documentVersionCache.TryGetDocumentVersion(documentSnapshot, out var version))
+                    {
+                        return new DocumentSnapshotAndVersion(documentSnapshot, version.Value);
+                    }
+                }
+
+                return null;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentAndVersion is null)
+            {
+                return null;
+            }
+
+            var (documentSnapshot, version) = documentAndVersion;
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
```

Comment:
Surprisingly this should actually improve perf because prior to this we'd be relying on calling `GetGeneratedOutputAsync` every time we needed a code document. This in turn could result in multiple parses happening for a single feature execution. That being said the `DocumentContext` should only ever be called if it's needed by a consuming feature which always needs the code document / syntax tree to do anything reasonable.

Reply:
Done: https://github.com/dotnet/razor-tooling/pull/6435/commits/6ae11db7277d6f8fcc06f15b929c9fa1b92d60a9